torch/distributed/fsdp/_init_utils.py

```diff
@@ -75,13 +88,140 @@
+        else:
+            state.process_group = process_group
+
     state.rank = state.process_group.rank()
     state.world_size = state.process_group.size()
+
     return state
 
+@no_type_check
+def _init_process_group_state_for_hybrid_shard(state, process_group) -> _FSDPState:
+    if process_group is None:
+        default_group = _get_default_group()
+        intra_node_group, inter_node_group = _init_intra_and_inter_node_groups(default_group)
+        # we shard across intra-node
+        state.process_group = intra_node_group
+        # save _inter_node_pg to allreduce across.
+        state._inter_node_pg = inter_node_group
+    else:
+        # Check type and assign state.process_group and state._inter_node_pg.
+        valid = _validate_hybrid_shard_pg_type(process_group)
+        if valid:
+            # Assuming that user passed in as intra node group and inter node group
+            # as documented.
+            state.process_group, state._inter_node_pg = process_group
+        else:
+            raise ValueError(
+                "Expected process_group to be passed in as either None or "
+                f"Tuple[dist.ProcessGroup, dist.ProcessGroup] but got {type(process_group)}"
+            )
+    # Create state for allreduce
+    state._inter_node_state = _get_default_comm_hook_state(
+        process_group=state._inter_node_pg,
+    )
+    return state
+
+@no_type_check
+def _validate_hybrid_shard_pg_type(process_group: Any) -> bool:
+    return (
+        isinstance(process_group, tuple)
+        and len(process_group) == 2
+        and all(isinstance(pg, dist.ProcessGroup) for pg in process_group)
+    )
+
+@no_type_check
+def _init_intra_node_process_group() -> dist.ProcessGroup:
+    """
+    Returns a process group across the current machine.
+    For example, given each row is a distinct machine:
+    0 3
+    1 4
+    2 5
+    This API would return an intra-node subgroup across
+    {0, 1, 2} or {3, 4, 5}, depending on the process's rank.
+    For example, rank 3 would get {3, 4, 5}.
+    """
+    intra_node_subgroup, _ = dist.new_subgroups()
+    return intra_node_subgroup
+
+@no_type_check
+def _init_inter_node_process_group(
+    global_process_group: dist.ProcessGroup
+) -> dist.ProcessGroup:
+    """
+    Returns an inter-node process group where each contained rank has
+    the same local rank. For example, given each row is a distinct machine:
+    0 3
+    1 4
+    2 5
+    This API would return inter-node process group {0, 3}, {1, 4}, or {2, 5},
+    depending on the process's rank. For example, ranks 1 and 4 would get {1, 4}.
+    """
+    # the inter-node pg that is returned
+    inter_node_pg = None
+    sharding_backend = dist.get_backend(global_process_group)
+    world_size = dist.get_world_size(global_process_group)
+    # Assuming fully homogeneous setup
+    num_devices = torch.cuda.device_count()
+    num_distinct_nodes = world_size // num_devices
```

Comment:
nit: Like my comment on the original PR,
```suggestion
    num_nodes = world_size // num_devices
```
(Does "distinct" mean something different than if it was not present?)

Reply:
[Easy] I recommend changing before landing